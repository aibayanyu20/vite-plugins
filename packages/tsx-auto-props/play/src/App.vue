<script setup lang="ts">
import Test from './components/Test'
</script>

<template>
  <div>
    <Test test="测试" test2="测试2" test4="sad111asds" />
  </div>
</template>
