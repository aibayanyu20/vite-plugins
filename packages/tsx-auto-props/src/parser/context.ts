export class Context {
  constructor() {
    this.mapFile = new Map()
  }

  mapFile: Map<string, string[]>
  add(fileName: string, id: string) {
    if (this.mapFile.has(fileName)) {
      const arr = this.mapFile.get(fileName)
      if (!arr?.includes(id))
        arr!.push(id)
    }
    else {
      this.mapFile.set(fileName, [id])
    }
  }

  get(fileName: string) {
    return this.mapFile.get(fileName)
  }

  getId(fileName: string, id: string) {
    const files = this.mapFile.get(fileName)
    if (files)
      return files.find(item => item === id)
  }

  has(fileName: string) {
    return this.mapFile.has(fileName)
  }

  hasId(fileName: string, id: string) {
    const files = this.mapFile.get(fileName)
    if (files)
      return files.includes(id)
    return false
  }
}
