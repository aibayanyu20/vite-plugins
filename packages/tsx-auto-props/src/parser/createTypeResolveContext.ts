import fs from 'node:fs'
import type { SimpleTypeResolveContext } from '@vue/compiler-sfc'
import { parse } from '@vue/compiler-sfc'
import type { Statement } from '@babel/types'
import type { Context } from './context'

export function createTypeResolveContext(code: string, ast: Statement[], id: string, context?: Context): SimpleTypeResolveContext {
  const setup = `<script lang="ts" setup>${code}</script>`
  const { descriptor } = parse(setup, {
    filename: id,
  })
  const helperImports: Set<string> = new Set()
  const error = (msg: string, _node: any, _scope?: any) => {
    // console.error(msg, node, scope)
    throw new Error(msg)
  }

  return {
    source: descriptor.source,
    filename: descriptor.filename,
    error,
    helper(key: string): string {
      helperImports.add(key)
      return `_${key}`
    },
    getString(node) {
      const block = descriptor.scriptSetup!
      return block.content.slice(node.start!, node.end!)
    },
    propsTypeDecl: undefined,
    propsRuntimeDefaults: undefined,
    propsDestructuredBindings: Object.create(null),
    emitsTypeDecl: undefined,
    ast,
    options: {
      fs: {
        fileExists(_file: string): boolean {
          // 检查文件是否存在
          try {
            const stat = fs.statSync(_file)
            if (stat.isFile()) {
              context?.add(_file, id)
              context?.addGraph(id, _file)
              return true
            }
            return false
          }
          catch (e) {
            return false
          }
        },
        readFile(_file: string): string | undefined {
          return fs.readFileSync(_file, 'utf-8')
        },
      },
    },
  }
}
