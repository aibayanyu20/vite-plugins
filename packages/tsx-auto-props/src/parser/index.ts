export { createTypeResolveContext } from './createTypeResolveContext'
export { createAst, traverse } from './createAst'
export { transform } from './transform'
