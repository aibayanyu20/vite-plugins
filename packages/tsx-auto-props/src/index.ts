import type { ModuleGraph, ModuleNode, PluginOption } from 'vite'
import { transform } from './parser'
import { Context } from './parser/context'
import { invalidateTypeCacheId } from './parser/transform'

function getDepModules(id: string, mapFile: Context, moduleGraph: ModuleGraph): ModuleNode[] {
  const modules: ModuleNode[] = []
  if (mapFile.has(id)) {
    invalidateTypeCacheId(id)
    const ids = mapFile.get(id)
    if (ids && ids.length) {
      for (const id of ids) {
        const mod = moduleGraph.getModuleById(id)
        if (mod) {
          modules.push(mod)
          modules.push(...getDepModules(id, mapFile, moduleGraph))
        }
      }
    }
  }
  const mod = moduleGraph.getModuleById(id)
  if (mod)
    modules.push(mod)
  return [...modules]
}

export interface TsxAutoPropsOptions {
  setup?: boolean
}

export function tsxAutoProps(options?: TsxAutoPropsOptions): PluginOption {
  const mapFile = new Context()
  return {
    name: 'tsx-auto-props',
    enforce: 'pre',
    transform(code, id) {
      if (id.endsWith('.tsx'))
        return transform(code, id, mapFile)
    },
    /**
     * 热更新
     */
    handleHotUpdate({ file, server, modules }) {
      if (mapFile.has(file)) {
        // 清理缓存
        invalidateTypeCacheId(file)
        // 如果存在那么响应的所有的moduleGraph都需要更新
        const ids = mapFile.get(file)
        const affectedModules = []
        if (ids) {
          for (const id of ids) {
            const mods = getDepModules(id, mapFile, server.moduleGraph)
            mapFile.removeCache(id)
            if (mods.length)
              affectedModules.push(...mods)
          }
          return [...modules, ...affectedModules]
        }
      }
      return modules
    },
  }
}
export default tsxAutoProps
