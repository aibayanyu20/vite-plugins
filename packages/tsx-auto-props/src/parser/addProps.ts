import * as t from '@babel/types'
import type { CallExpression } from '@babel/types'
import type { Parsed } from './typing'
import { traverse } from './createAst'
import { generateCode } from './utils'

function genCode(node: t.Node | t.File) {
  const res = generateCode(node)
  return res.code
}

/**
 * 检查当前的defineComponent的第一个参数是不是一个函数
 */
export function checkFunctionArgument(node: t.CallExpression, props: t.ArrayExpression) {
  const [arg] = node.arguments
  if (!t.isArrowFunctionExpression(arg) && !t.isFunctionExpression(arg))
    return false
    /**
     * 如果是就开始对第二个参数进行检查，如果存在第二个参数，那么我们就直接在第二个参数中进行添加props，如果不存在第二个参数，
     * 那么我们就要生成一个对象，并将props添加到对象中，然后将对象作为第二个参数传入
     */
  const arg2 = node.arguments[1]
  if (arg2) {
    // 存在第二个参数的情况
    if (t.isObjectExpression(arg2)) {
      // 如果第二个参数是一个对象，那么我们就直接在这个对象中添加一个props属性
      // 判断原始是不是存在一个props属性，如果存在就不添加了
      const propsProperty = arg2.properties.find(p => t.isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
      if (propsProperty)
        return false
      arg2.properties.push(t.objectProperty(t.identifier('props'), props))
    }
  }
  else {
    // 不存在第二个参数的情况
    // 生成一个对象
    const obj = t.objectExpression([t.objectProperty(t.identifier('props'), props)])
    // 将对象作为第二个参数传入
    node.arguments.push(obj)
  }
  return true
}

export function checkObjectArgument(node: t.CallExpression, props: t.ArrayExpression) {
  /**
   * 检查当前是不是一个对象，如果是一个对象,那么我们就直接在这个对象中添加一个props属性
   */
  const arg = node.arguments[0]
  if (t.isObjectExpression(arg)) {
    // 如果第二个参数是一个对象，那么我们就直接在这个对象中添加一个props属性
    // 判断原始是不是存在一个props属性，如果存在就不添加了
    const propsProperty = arg.properties.find(p => t.isObjectProperty(p) && 'name' in p.key && p.key.name === 'props')
    if (propsProperty)
      return false
    arg.properties.unshift(t.objectProperty(t.identifier('props'), props))
    return true
  }
  return false
}

/**
 * Add props to a defineComponent
 * @param parsed
 * @param props
 */
export function addProps(parsed: Parsed, props: string[]) {
  let code: string = ''
  const propsArray = t.arrayExpression(props.map(p => t.stringLiteral(p)))
  traverse(parsed, {
    ExportDefaultDeclaration({ node }) {
      const { declaration } = node
      if (t.isCallExpression(declaration) && t.isIdentifier(declaration?.callee) && declaration.callee?.name === 'defineComponent') {
        if (checkFunctionArgument(declaration, propsArray))
          code = genCode(node)
        else if (checkObjectArgument(declaration, propsArray))
          code = genCode(node)
      }
    },
    AssignmentExpression({ node }) {
      const right = node.right as CallExpression
      if (t.isCallExpression(right) && t.isIdentifier(right.callee) && right.callee.name === 'defineComponent') {
        if (checkFunctionArgument(right, propsArray))
          code = genCode(node)

        else if (checkObjectArgument(right, propsArray))
          code = genCode(node)
      }
    },
  })
  return code
}
