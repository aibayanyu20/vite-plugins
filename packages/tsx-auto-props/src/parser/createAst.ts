import { parse } from '@babel/parser'
import _traverse from '@babel/traverse'
import type { Parsed } from './typing'

export const traverse = typeof (_traverse as any).default === 'undefined'
  ? _traverse
  : ((_traverse as any).default as typeof _traverse)
export function createAst(code: string, jsx = true): Parsed {
  if (!jsx) {
    return parse(code, {
      sourceType: 'module',
      plugins: ['typescript'],
    })
  }
  return parse(code, {
    sourceType: 'module',
    plugins: ['typescript', 'jsx'],
  })
}
