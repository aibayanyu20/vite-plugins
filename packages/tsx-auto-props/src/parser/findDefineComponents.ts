import { isCallExpression, isIdentifier } from '@babel/types'
import type { Parsed } from './typing'
import { traverse } from './createAst'
import { findPropTypeMemberKeys } from './findPropTypeMemberKeys'

export type FindDefineComponentsResult = [ string | undefined, string, number, number]

export function findDefineComponents(parsed: Parsed): FindDefineComponentsResult[] {
  const result: FindDefineComponentsResult[] = []
  traverse(parsed, {
    VariableDeclarator({ node: { id, init, end, start } }) {
      if (isCallExpression(init) && isIdentifier(init?.callee) && isIdentifier(id) && init.callee?.name === 'defineComponent') {
        const memberKeys = findPropTypeMemberKeys(parsed, init)
        if (memberKeys)
          result.push([id.name, memberKeys, start!, end!])
      }
    },
    ExportDefaultDeclaration({ node: { declaration, end, start } }) {
      if (isCallExpression(declaration) && isIdentifier(declaration?.callee) && declaration.callee?.name === 'defineComponent') {
        const memberKeys = findPropTypeMemberKeys(parsed, declaration)
        if (memberKeys)
          result.push([undefined, memberKeys, start!, end!])
      }
    },
  })
  return result
}
