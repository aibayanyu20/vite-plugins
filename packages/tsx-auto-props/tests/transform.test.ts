import { fileURLToPath } from 'node:url'
import path from 'node:path'
import { describe, expect, it } from 'vitest'
import ts from 'typescript'
import { registerTS } from '@vue/compiler-sfc'
import { transform } from '../src/parser'
import singleRaw from './fixtures/single.tsx?raw'
import complexRaw from './fixtures/complex.tsx?raw'
import exportRaw from './fixtures/export.tsx?raw'
import nonTypeRaw from './fixtures/nonType.tsx?raw'
import inlineRaw from './fixtures/inline.tsx?raw'
import genericRaw from './fixtures/generic.tsx?raw'
import libRaw from './fixtures/lib.tsx?raw'
import renderRaw from './fixtures/render.tsx?raw'

registerTS(() => ts)

const fixturePath = fileURLToPath(new URL('./fixtures', import.meta.url))

describe('transform', () => {
  it('should transform', () => {
    const code = singleRaw
    const id = path.resolve(fixturePath, 'single.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })
  it('should complex', () => {
    const code = complexRaw
    const id = path.resolve(fixturePath, 'complex.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should export default', () => {
    const code = exportRaw
    const id = path.resolve(fixturePath, 'export.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should nonType', () => {
    const code = nonTypeRaw
    const id = path.resolve(fixturePath, 'nonType.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should inline', () => {
    const code = inlineRaw
    const id = path.resolve(fixturePath, 'inline.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should generic', () => {
    const code = genericRaw
    const id = path.resolve(fixturePath, 'generic.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should lib', () => {
    const code = libRaw
    const id = path.resolve(fixturePath, 'lib.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })

  it('should render', () => {
    const code = renderRaw
    const id = path.resolve(fixturePath, 'render.tsx')
    const transformCode = transform(code, id)
    expect(transformCode).toMatchSnapshot()
  })
})
