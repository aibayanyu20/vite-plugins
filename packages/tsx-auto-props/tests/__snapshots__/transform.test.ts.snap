// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform > should complex 1`] = `
{
  "code": "import { defineComponent } from 'vue'

export interface CommonProps {
  b: string
}

export interface ComplexProps {
  a: string
}

export const Complex = defineComponent<CommonProps & ComplexProps>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"b\\", \\"a\\"]
})

export const Complex1 = defineComponent((_props: CommonProps & ComplexProps) => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"b\\", \\"a\\"]
})

export const Complex2 = defineComponent((_props: CommonProps & {
  c: string;
}) => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"b\\", \\"c\\"]
})

export const Complex3 = defineComponent({
  props: [\\"b\\", \\"a\\"],
  setup(_props: CommonProps & ComplexProps) {
    return () => {
      return <div></div>;
    };
  }
})

export const Complex4 = defineComponent({
  props: ['a', 'b'],
  setup(_props: CommonProps & ComplexProps) {
    return () => {
      return <div></div>
    }
  },
})
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAMX;AACF;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should export default 1`] = `
{
  "code": "import { defineComponent } from 'vue'

export interface CommonProps {
  b: string
}

export default defineComponent<CommonProps>({
  props: [\\"b\\"],
  setup() {
    return () => {
      return <div></div>;
    };
  }
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,GAME;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should generic 1`] = `
{
  "code": "import { defineComponent, ref } from 'vue'
import type { Props } from './typing'

export const Generic = defineComponent(<T extends string | number,>(props: {
  msg: T;
  list: T[];
}) => {
  const count = ref(0);
  return () => <div>
          {props.msg}
          {' '}
          {count.value}
        </div>;
}, {
  props: [\\"msg\\", \\"list\\"]
})
export const Generic1 = defineComponent(<T extends Props | number,>(props: {
  msg: T;
  list: T[];
}) => {
  const count = ref(0);
  return () => <div>
          {props.msg}
          {' '}
          {count.value}
        </div>;
}, {
  props: [\\"msg\\", \\"list\\"]
})
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAYZ;AACD,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAYZ;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should inline 1`] = `
{
  "code": "import { defineComponent } from 'vue'

export const Inline = defineComponent((_props: {
  c: string;
  a: number;
}) => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"c\\", \\"a\\"]
})

export const Inline1 = defineComponent<{
  c: string;
  a: number;
}>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"c\\", \\"a\\"]
})
export const Inline2 = defineComponent({
  props: [\\"c\\", \\"a\\"],
  setup(_props: {
    c: string;
    a: number;
  }) {
    return () => {
      return <div></div>;
    };
  }
})

export default defineComponent({
  props: [\\"c\\", \\"a\\", \\"b\\"],
  setup(_props: {
    c: string;
    a: number;
    b: string;
  }) {
    return () => {
      return <div></div>;
    };
  }
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAMX;AACF;AACA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,GAME;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should lib 1`] = `
{
  "code": "import type { TableProps } from 'ant-design-vue'
import { defineComponent } from 'vue'

export type MyTableProps = TableProps & {
  sss?: string
}

export const LibTest = defineComponent<MyTableProps>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"dropdownPrefixCls\\", \\"dataSource\\", \\"columns\\", \\"pagination\\", \\"loading\\", \\"size\\", \\"bordered\\", \\"locale\\", \\"onChange\\", \\"onResizeColumn\\", \\"rowSelection\\", \\"getPopupContainer\\", \\"scroll\\", \\"sortDirections\\", \\"showSorterTooltip\\", \\"prefixCls\\", \\"rowKey\\", \\"tableLayout\\", \\"rowClassName\\", \\"title\\", \\"footer\\", \\"id\\", \\"showHeader\\", \\"components\\", \\"customRow\\", \\"customHeaderRow\\", \\"direction\\", \\"expandFixed\\", \\"expandColumnWidth\\", \\"expandedRowKeys\\", \\"defaultExpandedRowKeys\\", \\"expandedRowRender\\", \\"expandRowByClick\\", \\"expandIcon\\", \\"onExpand\\", \\"onExpandedRowsChange\\", \\"defaultExpandAllRows\\", \\"indentSize\\", \\"expandIconColumnIndex\\", \\"showExpandColumn\\", \\"expandedRowClassName\\", \\"childrenColumnName\\", \\"rowExpandable\\", \\"sticky\\", \\"transformCellText\\", \\"sss\\"]
})
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should nonType 1`] = `
{
  "code": "import { defineComponent } from 'vue'

export const NonType = defineComponent(() => {
  return () => {
    return <div></div>
  }
})
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`transform > should transform 1`] = `
{
  "code": "import { defineComponent } from 'vue'
import type { Props } from './typing'

export interface Props1 {
  foo1: string
  bar1: number
}

export type Props2 = Props & Props1
export type Props3 = Pick<Props2, 'foo1' | 'foo'>

export const Single = defineComponent<Props>(() => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"foo\\", \\"bar\\"]
})

export const Single2 = defineComponent((_props: Props2) => {
  return () => {
    return <div></div>;
  };
}, {
  props: [\\"foo\\", \\"bar\\", \\"foo1\\", \\"bar1\\"]
})

export const Single3 = defineComponent({
  props: [\\"foo1\\", \\"foo\\"],
  setup(_props: Props3) {
    return () => {
      return <div></div>;
    };
  }
})
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAIX;AACF;AACA,aAAa;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA;AAAA,EAMX;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
